proposals/010-workflow-migrations.md

```diff
@@ -0,0 +1,303 @@
+# Proposal: Workflow Migrations
+
+Author: Jake Rose
+
+This proposal will introduce the following:
+- `WfRunMigrationPlans`(metadata object)
+- `WfRunMigration` (run time action)
+- edge cases while migrating `wfRun`
+- Future Plans
+
+## Background
+
+Long-running processes are one of the main attractions of any durable execution engine, allowing state to survive for days, weeks, or even months. As a result, migrations become an important feature, since services and processes inevitably change over time.
+
+Migrations -- the practice of transferring runtime state from one actively running process (`wfRun`) to another compatible process specification (`wfSpec`) becomes a must have capability.
+
+
+## `WfRunMigrationPlan`
+
+`wfRunMigrationPlan` is a metadata object that sets the stage for how a migration will occur within `wfRun`
+
+You can think of the `WfRunMigrationPlan` as a set of instructions that will be enforced during a given migration.
+
+The rpc to register this metadata object in the lh server will look as follows:
+
+
+## Option 1
+### Proto changes
+```
+  rpc PutMigrationPlan(PutMigrationPlanRequest) returns (WfRunMigrationPlan) {}
+```
+
+After registering the metadata object, the lh server will return the object itself.
+
+The following is the wfRunMigrationPlan proto:
+
+```
+message WfRunMigrationPlan {
+    // Name
+    MigrationPlanId id = 1;
+
+    // When the MigrationPlan was created.
+    google.protobuf.Timestamp created_at = 2; 
+
+    // Wrapper object that stores how to migrate every thread
+    MigrationPlan migration_plan = 3 ;
+}
+```
+The first two fields are self-explanatory: the ID used to identify the plan and the time it was created. The third field is a wrapper class where metadata about the migration begins.
+
+The `MigrationPlan` consists of the following proto:
+
+```
+message MigrationPlan { 
+    // {Threadname: {How to migrate that thread}}
+    map<string, ThreadMigrationPlan> thread_migrations = 1;
+}
+```
+Migration plans are broken down to a thread by thread basis.
```

Comment:
It would be helpful in this proposal if you specify between `ThreadSpec` vs `ThreadRun` :slightly_smiling_face: 

There are some nuances because a single `ThreadSpec` could have zero or one or ten corresponding `ThreadRun`s.